project/src/com/google/daggerquery/executor/models/GraphProto.java

```diff
@@ -29,9 +33,30 @@
 public class GraphProto implements Graph {
 
   private BindingGraph bindingGraph;
+  private ImmutableMap<String, ImmutableSet<String>> reversedBindingGraph;
 
   public GraphProto(BindingGraph bindingGraph) {
     this.bindingGraph = bindingGraph;
+
+    Map<String, ImmutableSet.Builder<String>> reversedGraphModel = new HashMap<>();
+
+    for (String source: bindingGraph.getAdjacencyListMap().keySet()) {
+      reversedGraphModel.put(source, new ImmutableSet.Builder<>());
+    }
+
+    for (String source: bindingGraph.getAdjacencyListMap().keySet()) {
+      for (Dependency dependency: bindingGraph.getAdjacencyListMap().get(source).getDependencyList()) {
+        String target = dependency.getTarget();
+        reversedGraphModel.get(target).add(source);
+      }
+    }
+
+    this.reversedBindingGraph = ImmutableMap.copyOf(
+      Maps.transformValues(
+        reversedGraphModel,
+        ImmutableSet.Builder::build
+      )
+    );
```

Comment:
nit: How about moving this logic to a helper method for clarity?

Reply:
[Done](https://github.com/googleinterns/dagger-query/pull/43/commits/d2943267efcf2e2238b051c76b286acaad265bb2) ✅